CmdlineRedactor.cpp

```diff
@@ -0,0 +1,113 @@
+*/
+
+#include "CmdlineRedactor.h"
+
+#include "Config.h"
+#include "FileUtils.h"
+#include "Logger.h"
+#include "StringUtils.h"
+
+std::shared_ptr<CmdlineRedactionRule> CmdlineRedactionRule::LoadFromFile(const std::string& path) {
+    Config config;
+    try {
+        config.Load(path);
+
+        std::string replacement_char = "*";
+        if (config.HasKey("replacement_char")) {
+            replacement_char = config.GetString("replacement_char");
+        }
+
+        if (!config.HasKey("regex")) {
+            Logger::Error("CmdlineRedactionRule::LoadFromFile(): Config (%s) is missing the 'regex' value", path.c_str());
+            return nullptr;
+        }
+        std::string regex = config.GetString("regex");
+
+        auto ret = std::make_shared<CmdlineRedactionRule>(regex, replacement_char[0]);
+        if (!ret->Compile()) {
+            Logger::Error("CmdlineRedactionRule::LoadFromFile(): Failed to load from (%s): Invalid regex: %s", path.c_str(), ret->CompileError().c_str());
+            return nullptr;
+        }
+        return ret;
+    } catch (std::exception& ex) {
+        Logger::Error("CmdlineRedactionRule::LoadFromFile(): Failed to load from (%s): %s", path.c_str(), ex.what());
+    }
+
+    return nullptr;
+}
+
+bool CmdlineRedactionRule::Check(std::string& cmdline) const {
+    if (cmdline.empty()) {
+        return false;
+    }
+    auto res = false;
+    size_t num_groups = _regex.NumberOfCapturingGroups() + 1;
+    re2::StringPiece groups[num_groups];
+    re2::StringPiece text(cmdline);
+    size_t idx = 0;
+    size_t end = text.size();
+    while (idx < end) {
+        if (!_regex.Match(text, idx, end, re2::RE2::UNANCHORED, &groups[0], num_groups)) {
+            break;
+        }
+        for (int i = 1; i < num_groups; ++i) {
+            if (groups[i].data() != nullptr) {
+                char * start = cmdline.data() + (groups[i].data() - text.data());
+                std::fill(start, start + groups[i].size(), _replacement_char);
+            }
+        }
+        res = true;
+        idx += groups[0].size();
+    }
+    return res;
+}
+
+void CmdlineRedactor::AddRule(const std::shared_ptr<CmdlineRedactionRule>& rule) {
+    _rules.emplace_back(rule);
+}
+
+void CmdlineRedactor::LoadFromDir(const std::string& dir) {
+    if (!PathExists(dir)) {
+        return;
+    }
+    std::vector<std::string> files;
+    try {
+        files = GetDirList(dir);
+    } catch (std::exception& ex) {
+        Logger::Error("CmdlineRedactor::LoadFromDir(): Failed to read dir (%s): %s", dir.c_str(), ex.what());
+        return;
+    }
+    std::sort(files.begin(), files.end());
+    for (auto& name: files) {
+        if (ends_with(name, ".conf")) {
+            auto rule = CmdlineRedactionRule::LoadFromFile(dir + "/" + name);
+            if (rule) {
+                _rules.emplace_back(rule);
+            }
+        }
+    }
+}
+
+bool CmdlineRedactor::Check(std::string& cmdline) const {
```

Comment:
>Check [](start = 22, length = 5)

As check() is updating/filling the given cmdline string after checking, consider renaming the function, may be rewrite, redact, or other suitable name?

Reply:
Renamed.